### ISSUE TITLE
Support column limit in formatter

### ISSUE THREAD

**Issue by alexeagle** [Microsoft/TypeScript#1988]
Currently, running the formatter
(eg via the compiler API: https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#pretty-printer-using-the-ls-formatter)
produces code which is arbitrarily wide, for statements which are nested, complex, etc.

Our style guide requires a maximum columns, which means the formatter should take a configuration for how wide this is.
Line breaks should probably be added at the highest point in the AST on the long line.

**Comment by danquirk** [comment]
Can you give an example of code you write that is not beyond your max column limit until the formatter takes action?

**Comment by mhegazy** [comment]
The formatter behavior tries to maintain lines as much as possible. this looks like a good idea, can you provide some details on the proposal, specially with ASI taking effect in certain cases.

**Comment by alexeagle** [comment]
Indeed, it does preserve the lines from the input. In my case, the input violates the style guide, and I'd like to use the formatter to fix that.

reminder about the context: I'm working in the Closure Compiler to insert TypeScript syntax, so in general most lines become longer. In bad cases, we use a union of several types like 
```javascript
goog.string.compareElements_ = function(left: string | number | boolean, right: string | number | boolean): number {
```

if we want to convert chunks of Google javascript code to this syntax, we need to be able to style the generated result.

I imagine this also is the case in IDE support, for example if you do a refactoring that inlines some code.

**Comment by mhegazy** [comment]
I agree that this is a good feature to add. The problem is you can not just add a new line after a token, as it may not be safe. e.g. after a return keyword, and before its expression. So from a cursory look, when formatting a token pair, if the line position surpasses the maximum allowed, we walk back starting with the current pair, and try to find a location where it is safe to break the line, assuming one exists, then add the edit. alternativelly, we can do that as a separate pass, after we are done formatting and indenting, and just break lines that are longger than the allowed range.
we need to get a more detailed proposal for this. @vladima, ideas?

**Comment by DanielRosenwasser** [comment]
[[Wadler, 98]](http://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf) has a motivation point for a column-width aware pretty-printer.

**Comment by alexeagle** [comment]
FYI we are also considering extending clang-format to properly format typescript code. It's already pretty close if you use JavaScript as the language. Dunno if it's useful as a reference for you.

```bash
$ brew install clang-format
$ git clone git@github.com:Microsoft/TypeScriptSamples.git
$ clang-format -style=google -assume-filename=foo.js TypeScriptSamples/greeter/greeter.ts
class Greeter {
  constructor(public greeting : string) {}
  greet() { return "<h1>" + this.greeting + "</h1>"; }
};
var greeter = new Greeter("Hello, world!");
var str = greeter.greet();
document.body.innerHTML = str;
```

**Comment by mhegazy** [comment]
PR would be welcomed for this one as well on the typescript side.
